### PR TITLE
Issue-1824: added flags and different auth types support

### PR DESCRIPTION
**Pull request by dmitryk-dk**
https://github.com/VictoriaMetrics/VictoriaMetrics/issues/1824

### CODE REVIEW

File: app/vmalert/notifier/init.go [hunk 1]
```diff
@@ -15,18 +15,37 @@ var (
...
-	addrs                 = flagutil.NewArray("notifier.url", "Prometheus alertmanager URL, e.g. http://127.0.0.1:9093")
+	addrs = flagutil.NewArray("notifier.url", "Prometheus alertmanager URL, e.g. http://127.0.0.1:9093")
+
 	basicAuthUsername     = flagutil.NewArray("notifier.basicAuth.username", "Optional basic auth username for -notifier.url")
 	basicAuthPassword     = flagutil.NewArray("notifier.basicAuth.password", "Optional basic auth password for -notifier.url")
 	basicAuthPasswordFile = flagutil.NewArray("notifier.basicAuth.passwordFile", "Optional path to basic auth password file for -notifier.url")
 
+	authorizationType            = flagutil.NewArray("notifier.authorization.type", "Optional authorization type (`basic or bearer`) for -notifier.url")
+	authorizationCredentials     = flagutil.NewArray("notifier.authorization.credentials", "Optional authorization credentials for -notifier.url")
+	authorizationCredentialsFile = flagutil.NewArray("notifier.authorization.credentialsFile", "Optional path to authorization credentials file for -notifier.url")
```
Comment (hagen1778) [inline app/vmalert/notifier/init.go:26]: I think we can ignore `authorization` options here and in the rest of packages from this PR.
Looks like we don't use it in vmagent as well, except of auth options configured via promscrape file. 
@f41gh7 is my understanding correct?

File: app/vmalert/datasource/vm_test.go [hunk 2]
```diff
@@ -150,6 +150,139 @@ func TestVMInstantQuery(t *testing.T) {
 	}
 }
 
+func TestVMInstantQueryWithAuthorization(t *testing.T) {
```
Comment (hagen1778) [inline app/vmalert/datasource/vm_test.go:153]: the test looks too excessive and contains a lot of copy-pasted code.
Testing the auth headers can be done in a more straightforward way of just checking the headers. Please, consider https://github.com/VictoriaMetrics/VictoriaMetrics/blob/ef1afeed6c2bd427aee92d15e78c1330926b5af2/app/vmalert/datasource/vm_test.go#L374-L382 as an example

File: app/vmalert/notifier/alertmanager_test.go [hunk 3]
```diff
@@ -112,3 +113,95 @@ func TestAlertManager_Send(t *testing.T) {
 		t.Errorf("expected 2 calls(count from zero) to server got %d", c)
 	}
 }
+
+func TestAlertManager_SendWithAuthorization(t *testing.T) {
```
Comment (hagen1778) [inline app/vmalert/notifier/alertmanager_test.go:117]: The same problem here as in previous test. Either extend original test function or come up with a simpler test which check auth params only.

**Review by hagen1778 — CHANGES_REQUESTED** [review]
Thanks! Please see my comments

File: app/vmalert/notifier/alertmanager_test.go (hunk 3 above)
Comment (dmitryk-dk) [reply app/vmalert/notifier/alertmanager_test.go]: removed, implemented extended test

File: app/vmalert/utils/auth.go [hunk 4]
```diff
@@ -1,18 +1,79 @@
...
+		}
+		if credentials != "" {
+			authCfg.Credentials = promauth.NewSecret(credentials)
+		}
+		if credentialsFile != "" {
+			authCfg.CredentialsFile = credentialsFile
+		}
+		if authCfg.Type != "" || authCfg.Credentials.String() != "" || authCfg.CredentialsFile != "" {
+			config.Authorization = authCfg
+		}
```
[the remark this answers is not in the document]
Comment (dmitryk-dk) [reply app/vmalert/utils/auth.go]: fixed

File: app/vmalert/datasource/vm_test.go (hunk 2 above)
Comment (dmitryk-dk) [reply app/vmalert/datasource/vm_test.go]: removed, implemented extended test

File: app/vmalert/notifier/init.go (hunk 1 above)
Comment (f41gh7) [reply app/vmalert/notifier/init.go:26]: Yep, i think there is no need in these options. Bearer token and basic auth should be fine.

Comment (dmitryk-dk) [reply app/vmalert/notifier/init.go:26]: removed

File: app/vmalert/datasource/vm_test.go [hunk 5]
```diff
@@ -517,6 +518,61 @@ func TestRequestParams(t *testing.T) {
 	}
 }
 
+func TestAuthConfig(t *testing.T) {
+	// c := &http.Client{}
```
Comment (hagen1778) [inline app/vmalert/datasource/vm_test.go:522]: nit: commented line

File: app/vmalert/datasource/vm_test.go [hunk 6]
```diff
@@ -517,6 +518,61 @@ func TestRequestParams(t *testing.T) {
...
+	var testCases = []struct {
+		name string
+		// queryRange bool
```
Comment (hagen1778) [inline app/vmalert/datasource/vm_test.go:525]: nit: commented line

File: app/vmalert/notifier/init.go [hunk 7]
```diff
@@ -15,18 +15,37 @@ var (
...
 	suppressDuplicateTargetErrors = flag.Bool("notifier.suppressDuplicateTargetErrors", false, "Whether to suppress 'duplicate target' errors during discovery")
 
-	addrs                 = flagutil.NewArray("notifier.url", "Prometheus alertmanager URL, e.g. http://127.0.0.1:9093")
+	addrs = flagutil.NewArray("notifier.url", "Prometheus alertmanager URL, e.g. http://127.0.0.1:9093")
+
 	basicAuthUsername     = flagutil.NewArray("notifier.basicAuth.username", "Optional basic auth username for -notifier.url")
 	basicAuthPassword     = flagutil.NewArray("notifier.basicAuth.password", "Optional basic auth password for -notifier.url")
 	basicAuthPasswordFile = flagutil.NewArray("notifier.basicAuth.passwordFile", "Optional path to basic auth password file for -notifier.url")
 
+	authorizationType            = flagutil.NewArray("notifier.authorization.type", "Optional authorization type (`basic or bearer`) for -notifier.url")
```
Comment (hagen1778) [inline app/vmalert/notifier/init.go:24]: I guess, we can remove `*. authorization.*` flags as was discussed previously

**Review by hagen1778 — CHANGES_REQUESTED** [review]

File: app/vmalert/datasource/vm_test.go (hunk 5 above)
Comment (dmitryk-dk) [reply app/vmalert/datasource/vm_test.go]: removed

File: app/vmalert/datasource/vm_test.go (hunk 6 above)
Comment (dmitryk-dk) [reply app/vmalert/datasource/vm_test.go]: removed

File: app/vmalert/notifier/init.go (hunk 7 above)
Comment (dmitryk-dk) [reply app/vmalert/notifier/init.go]: removed

**Review by hagen1778 — APPROVED** [review]
LGTM!

**Comment by hagen1778** [comment]
@dmitryk-dk would you mind to copy new flags description to https://docs.victoriametrics.com/vmalert.html#flags ?

**Review by f41gh7 — APPROVED** [review]
LGTM

File: docs/vmalert.md [hunk 8]
```diff
@@ -89,20 +102,21 @@ groups:
...
 
 Each group has the following attributes:
+
 ```yaml
 # The name of the group. Must be unique within a file.
 name: <string>
 
-# How often rules in the group are evaluated.
-[ interval: <duration> | default = -evaluationInterval flag ]
+  # How often rules in the group are evaluated.
```
Comment (hagen1778) [inline docs/vmalert.md:110]: looks like some auto-formatter broke the indentations in the README file. Would you mind to revert it and apply changes only to flags section?

**Review by hagen1778 — CHANGES_REQUESTED** [review]

File: docs/vmalert.md (hunk 8 above)
Comment (dmitryk-dk) [reply docs/vmalert.md]: sorry something went wrong with formatting, fixed

**Review by hagen1778 — APPROVED** [review]

**Pull request merged**